sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -863,6 +868,36 @@ class Analyzer(override val catalogManager: CatalogManager)
     }
   }
 
+  object ResolveMelt extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+      _.containsPattern(MELT), ruleId) {
+
+      // once children and ids are resolved, we can determine values, if non were given
+      case m: Melt if m.childrenResolved && m.ids.forall(_.resolved) && m.values.isEmpty =>
+        m.copy(values = m.child.output.diff(m.ids))
```

Comment:
if the id is non-attribute `a + b`, then all attributes are value columns?

Reply:
And if the id expression is `$"a".as("my_name")`, then `$"id"` will also be a value expression?

Reply:
Now I got it:
```
df.unpivot(Array($"id" * 2), Array.empty, "var", "val")
df.unpivot(Array($"id".as("uid")), Array.empty, "var", "val")
```
Yes, all columns will be unpivoted (become values), including the "id" column.